oprf/group/suite.go

```diff
@@ -0,0 +1,201 @@
+	return &Scalar{c, c.curve.Params().N}
+}
+
+func getH2CSuite(c *Ciphersuite) (HashToPoint, error) {
+	var suite h2c.SuiteID
+
+	switch c.Name() {
+	case "OPRFP256-SHA512-ELL2-RO":
+		suite = h2c.P256_XMDSHA256_SSWU_RO_
+	default:
+		return nil, errors.New("invalid suite")
+	}
+
+	hasher, err := suite.Get([]byte(c.dst))
+	if err != nil {
+		return nil, err
+	}
+
+	return eccHasher{c, hasher}, nil
+}
+
+// HashToPoint produces a new point by hashing the input message.
+type HashToPoint interface {
+	Hash(msg []byte) (*Point, error)
+}
+
+type eccHasher struct {
+	suite  *Ciphersuite
+	hasher h2c.HashToPoint
+}
+
+func (h eccHasher) Hash(in []byte) (*Point, error) {
+	q := h.hasher.Hash(in)
+	x := q.X().Polynomial()
+	y := q.Y().Polynomial()
+
+	p := &Point{h.suite, new(big.Int), new(big.Int)}
+	p.x.Set(x[0])
+	p.y.Set(y[0])
+
+	// TODO: failing here
+	//if !p.IsValid() {
+	//	return nil, errors.New("invalid point")
+	//}
+
+	return p, nil
+}
+
+// HashToGroup performs a transformation to encode bytes as a Element object in the
+// group.
+func (c *Ciphersuite) HashToGroup(in []byte) (*Point, error) {
+	hasher, err := getH2CSuite(c)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := hasher.Hash(in)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// HashToScalar performs a transformation to encode bytes as a Scalar object in the
+// appropiate group.
+func (c *Ciphersuite) HashToScalar([]byte) (*Scalar, error) {
+	return &Scalar{}, nil
+}
+
+// RandomScalar samples a random scalar value from the field of scalars defined by the
+// group order.
+// TODO: not constant time
+func (c *Ciphersuite) RandomScalar() (*Scalar, error) {
+	N := c.Order()
+	bitLen := N.X.BitLen()
+	byteLen := (bitLen + 7) >> 3
+	buf := make([]byte, byteLen)
+
+	// rejection sampling
+	for {
+		_, err := io.ReadFull(rand.Reader, buf)
+		if err != nil {
+			return nil, errors.New("scalar generation failed")
+		}
+		// Mask to account for field sizes that are not a whole number of bytes.
+		var mask = []byte{0xff, 0x1, 0x3, 0x7, 0xf, 0x1f, 0x3f, 0x7f}
+		buf[0] = buf[0] & mask[bitLen%8]
+
+		// Check if scalar is in the correct range.
+		if new(big.Int).SetBytes(buf).Cmp(N.X) >= 0 {
+			continue
+		}
+		break
+	}
+
+	x := new(big.Int).SetBytes(buf)
+	return &Scalar{c, x}, nil
+}
+
+// ScalarMultBase performs a scalar multiplication of the Generator with some scalar
+// input
+func (c *Ciphersuite) ScalarMultBase(s *Scalar) (*Point, error) {
+	gen := c.Generator()
+
+	return gen.ScalarMult(s)
+}
+
+// ByteLength returns the ByteLength of objects associated with the Ciphersuite.
+func (c *Ciphersuite) ByteLength() int {
+	return (c.curve.Params().BitSize + 7) / 8
+}
+
+// NewSuite creates a new ciphersuite for the requested name.
+func NewSuite(name string) (*Ciphersuite, error) {
+	cSuite := &Ciphersuite{}
+
+	switch name {
+	case "P-256":
+		cSuite.name = "OPRFP256-SHA512-ELL2-RO"
+		cSuite.dst = "RFCXXXX-P256_XMD:SHA-512_SSWU_RO_"
```

Comment:
I think there should be no trailing "_". Per the draft, the DST is "RFCXXXX-P256XMD:SHA-256_SSWU_RO".

Reply:
(The Sage code is wrong here!)

Reply:
Oh! I'll change. Maybe tomorrow I'll change some parts of the sage code as well, that I found.

Reply:
Mmm.. so I find that the dst in the draft is: `RFCXXXX-P256_XMD:SHA-256_SSWU_RO_` with the trailing.. or maybe it is a different draft version?

Reply:
This should be solved now.